### ISSUE TITLE
Allow to specify the user when running a container

### ISSUE THREAD

**Issue by hatchan** [wercker/wercker#229]
Some containers change the user that is running inside the container. If this is a non-privileged user, then wercker is not able to create any directories on `/`.

**Comment by kokaz** [comment]
@hatchan Do you have more details on this issue?

**Comment by hatchan** [comment]
You can use any docker image as the base image. Some images override the user so potentially it is a non root users. This non root user is unable to create the directories that wercker requires.

So in the wercker.yml the user should be able to set the user name or user id which will be used during the wercker build.

**Comment by kokaz** [comment]
Is #28 related somehow?

**Comment by hatchan** [comment]
@kokaz yes it is

**Comment by sekka1** [comment]
At Dockercon 2017 security was one of the big topics.  This session went into a bunch of CVEs and why was there a vulnerability and how was it fixed.  They still havent published the youtube of this session but one of things he said was that a lot of these CVEs wouldnt be exploitable if you were not running as the "root" user.

https://dockercon.smarteventscloud.com/connect/sessionDetail.ww?SESSION_ID=108529

Running as root at the build stage is probably a must have but I think we should really  encourage our users to not run as a root user on the image that they push into their registry and run in production.

**Comment by Spy-Seth** [comment]
I'm having the same issue with some custom PHP image. I run the container with dedicated PHP user and it failed to create the `/pipeline` directory. I've try to prepare this directory when I build the image with the good right on him but now it try to remove it `rm -rf /pipeline`... 

Did you have any idea/feedback on this?

**Comment by vladak** [comment]
In the process of trying to run OpenGrok build also on Wercker (it is not really a substitute to Travis since Travis can run the builds on macOS, and also seamlessly create Github releases), I was baffled with this piece of Java code failing:

```java
        File temp = Files.createTempDirectory("opengrok-src-root4").toFile();
        temp.deleteOnExit();
        temp.setReadable(false);
        assertTrue(temp.isDirectory());
        assertTrue(temp.exists());
        assertFalse(temp.canRead());
```

while the last assertion should fail, it did not. Sure, enough I found that the test runs under UID 0. This was not happening in Travis since it runs under non-root user. Now I will have to go and implement some ugly annotation code which checks the UID so the unit test method can be skipped.

**Comment by vladak** [comment]
Here's workaround we are using:

```
build:                                                                          
  steps:                                                                        
    - install-packages:                                                         
      packages: git subversion mercurial sudo maven pep8                        
    - script:                                                                   
      name: Maven install                                                       
      code: |                                                                   
        /usr/sbin/groupadd -g 1000 foobar                                       
        /usr/sbin/useradd -u 1000 -g 1000 -s /bin/bash -m foobar                
        echo $WERCKER_ROOT > /tmp/wroot                                         
        chown -R foobar $WERCKER_ROOT                                           
        /usr/bin/sudo -i -u foobar -- bash -c 'cd `cat /tmp/wroot`; /usr/bin/mvn verify'
```

**Comment by vladak** [comment]
The `wercker/maven` was updated by @markxnelson to support non-root user however this is just one of the pieces and one still has to create the user (and install sudo !) beforehand.

**Comment by markxnelson** [comment]
Well, that depends on your choice of box.  If you choose to use a box without sudo in it, you have to install it.

**Comment by vladak** [comment]
Yeah, the comment was sort of nagging to see if the architecture could somehow figure the requirement and install sudo (and the user possibly) automatically.

**Comment by markxnelson** [comment]
Well we could try to, but it would be pretty challenging to make sure it worked for every possible box a user might happen to use, since you can choose any docker image at all... so it seems like a reasonable decision to expect the user to do it.  Maybe we could implement something that tries to do the right thing, at least for the majority of common cases...
